include/hipipe/core/stream/column.hpp

```diff
@@ -54,44 +106,90 @@ class column_base {
       : value_{rhs}
     {}
 
-    // conversion operators //
+    // value accessors //
+
+    std::vector<T>& value() { return value_; }
+    const std::vector<T>& value() const { return value_; }
+};
+
+
+class batch_t {
+private:
+    std::unordered_map<std::type_index, std::unique_ptr<abstract_column>> columns_;
 
-    operator std::vector<T>&() &
+public:
+    // value extraction //
+
+    template<typename Column>
+    void throw_check_extraction_type() const
     {
-        return value_;
+        if (columns_.count(std::type_index{typeid(Column)}) == 0) {
+            throw std::runtime_error{
+              std::string{"Trying to retrieve column `"} + typeid(Column).name()
+              + "`, but the batch contains no such column."};
+        }
     }
 
-    operator std::vector<T>&&() &&
+    template<typename Column>
+    typename Column::batch_type& extract()
     {
-        return std::move(value_);
+        throw_check_extraction_type<Column>();
+        return columns_.at(std::type_index{typeid(Column)})->extract<Column>();
     }
 
-    // value accessors //
+    template<typename Column>
+    const typename Column::batch_type& extract() const
+    {
+        throw_check_extraction_type<Column>();
+        return columns_.at(std::type_index{typeid(Column)})->extract<Column>();
+    }
 
-    std::vector<T>& value() { return value_; }
-    const std::vector<T>& value() const { return value_; }
-};
+    // raw column access //
 
-/// Specialization of column_base for non-copy-constructible types.
-template <typename T>
-struct column_base<T, false> : column_base<T, true> {
-    using column_base<T, true>::column_base;
+    template<typename Column>
+    Column& at()
+    {
+        throw_check_extraction_type<Column>();
+        return *columns_.at(std::type_index{typeid(Column)});
```

Comment:
why not `columns_->at` ?

Reply:
That would be a different thing. `columns_` is not a pointer, it is a `map`. I want to dereference the pointer retrieved by `at()`. Operator `*` has a lower priority than `.`.